patches/api/0475-WIP-DataKey-API.patch

```diff
@@ -0,0 +1,560 @@
++
++import io.papermc.paper.datakey.DataKey;
++import org.jetbrains.annotations.Nullable;
++import java.util.Set;
++
++public interface DataKeyMap {
++
++    static DataKeyMap empty() {
++        return DataKeyMapBridge.Holder.BRIDGE.empty();
++    }
++
++    @Nullable
++    <T> T get(DataKey.Valued<T> type);
++
++    Set<DataKey> keySet();
++
++    boolean has(DataKey type);
++
++    default <T> T getOrDefault(DataKey.Valued<T> type, T fallback) {
++        T object = this.get(type);
++        return object != null ? object : fallback;
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/datakey/map/DataKeyMapBridge.java b/src/main/java/io/papermc/paper/datakey/map/DataKeyMapBridge.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..9da6ab689eb43eb960429fe9a752c0c6c4183232
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datakey/map/DataKeyMapBridge.java
+@@ -0,0 +1,26 @@
++package io.papermc.paper.datakey.map;
++
++import net.kyori.adventure.util.Services;
++import org.bukkit.Material;
++import org.bukkit.inventory.meta.ItemMeta;
++import org.jetbrains.annotations.ApiStatus;
++
++@ApiStatus.Internal
++public interface DataKeyMapBridge {
++
++    PatchedDataKeyMap of(DataKeyMap keyMap);
++
++    DataKeyMap empty();
++
++    DataKeyMap itemStack(Material material);
++
++    PatchedDataKeyMap meta(Material material, ItemMeta itemMeta);
++
++    ItemMeta toMeta(Material material, PatchedDataKeyMap dataKeyMap);
++
++    class Holder {
++
++        public static DataKeyMapBridge BRIDGE = Services.service(DataKeyMapBridge.class)
++            .orElseThrow();
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/datakey/map/PatchedDataKeyMap.java b/src/main/java/io/papermc/paper/datakey/map/PatchedDataKeyMap.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..3340e13d6f161c521af5bca7a57c173bc0bdfead
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datakey/map/PatchedDataKeyMap.java
+@@ -0,0 +1,36 @@
++package io.papermc.paper.datakey.map;
++
++import io.papermc.paper.datakey.DataKey;
++import io.papermc.paper.datakey.patch.DataKeyMapPatch;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++public interface PatchedDataKeyMap extends DataKeyMap {
++
++    static PatchedDataKeyMap empty() {
++        return DataKeyMapBridge.Holder.BRIDGE.of(DataKeyMapBridge.Holder.BRIDGE.empty());
++    }
++
++    static PatchedDataKeyMap of(DataKeyMap map) {
++        return DataKeyMapBridge.Holder.BRIDGE.of(map);
++    }
++
++    @Nullable
++    <T> T set(DataKey.Valued<T> type, @Nullable T value);
++
++    // Returns if this value was previously set
++    boolean set(DataKey.NonValued type);
++
++    @Nullable
++    <T> T remove(DataKey.Valued<T> type);
++
++    // Returns if this value was previously set
++    boolean remove(DataKey.NonValued type);
++
++    @NotNull
++    DataKeyMapPatch asPatch();
++
++    void applyPatch(@NotNull DataKeyMapPatch patch);
```

Comment:
This should just accept/return PatchedDataKeyMap directly and possibly be moved to ItemStack or a separate interface

Reply:
Do we not want some kind of middle man object? So for example people could apply certain changes to multiple item stacks at once? (I would allow people to make their own patch sets)